### DISCUSSION TITLE
Is fief the right tool for me?

### DISCUSSION THREAD

**Discussion by meteoDaniel** [fief-dev/fief#372] [Q&A]
Dear Fief-Team, 

I really like your tool so far, but I am still asking myself whether it is the right tool or not. So I hope you can answer my questions. 
Before I want to say, that I have very limited knowledge in State-Of-The-Art User management. 

Current: 
I have a FastAPI as a REST interface to customer data. So a very common setup. Currently I am using AWS-Lambda + API Gateway for managing Access Tokens. 

Questions: 
1. Is it possible that a customer registerted within a Fief tenant can receive an Access Token for the FastAPI? Especially fully programatically, so without typing in username and password to the front-end? 
2. Am I right that OAuth2 means that an access token has to requested e.g. each time the customer want to access the api? In comparison, API Gateway creates One Token for lifetime. 

These are my questions so far. Really looking forward to your answers. 
Thanks 
Daniel

**Comment by fief-bailiff** [comment]
Hail, @meteoDaniel 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to [star our repository](https://github.com/fief-dev/fief) ⭐️

<p align="center">
<img width="656" alt="star-fief" src="https://github.com/fief-dev/fief/assets/1144727/b2db7f59-102f-47e5-9751-67fa10372c52">
</p>

## Want to support us?

Subscribe to one of our paid plan to help us continue our work and receive exclusive information and benefits! Starts at $5/month 🪙

<p align="center">
    <a href="https://polar.sh/fief-dev">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/embed/subscribe.svg?org=fief-dev&darkmode=1">
      <img alt="Subscribe" src="https://polar.sh/embed/subscribe.svg?org=fief-dev">
    </picture>
    </a>
</p>

Farewell!

**Comment by frankie567** [comment]
Hi @meteoDaniel 👋 

If I understand correctly, what you need is a way to issue a token so your users can query their data programmatically from your API, right?

Fief can help if you want to opt for the OAuth2 protocol. It's a bit more involved for your users since you'll have to generate a [Client](https://docs.fief.dev/configure/clients/) for them, and then they'll have to go through the OAuth2 authorization mechanism to get an access token. As you mentioned, this token has a short lifespan, but Fief supports *refresh tokens* which allows to regenerate a fresh access token without reauthorizing again.

This kind of setup works well for an application (e.g. a web frontend or mobile app) or if your users want to actually integrate your app into theirs.

---

If your need is "just" to allow your users to make queries quickly with a token generated once (commonly called a *Personal Access Token*), then you'll have to come with your own solution. It can be quite simple: for example, those tokens can be stored in your database. When a request comes in, you can check in the headers if it's there and if it matches a token in your DB.

**Reply by meteoDaniel** [reply]
Thanks a lot for your answer. 
For sure it is quite simple now. I will use fief to grant access to a web based portal. After successfully logged into the portal the user can retrieve a token for the API. The token directly comes from API gateway. 

Thanks for helping me out !
